src/runtime/interface.go

```diff
@@ -5,25 +5,69 @@ package runtime
 // Interfaces are represented as a pair of {typecode, value}, where value can be
 // anything (including non-pointers).
 
-import "unsafe"
+import (
+	"reflect"
+	"unsafe"
+)
 
 type _interface struct {
 	typecode uintptr
 	value    unsafe.Pointer
 }
 
 // Return true iff both interfaces are equal.
-func interfaceEqual(x, y _interface) bool {
-	if x.typecode != y.typecode {
-		// Different dynamic type so always unequal.
+func interfaceEqual(x, y interface{}) bool {
+	return shallowEqual(reflect.ValueOf(x), reflect.ValueOf(y))
+}
+
+func shallowEqual(x, y reflect.Value) bool {
+	if x.Type() == 0 || y.Type() == 0 {
+		return x.Type() == y.Type()
+	}
+
+	if x.Type() != y.Type() {
 		return false
 	}
-	if x.typecode == 0 {
-		// Both interfaces are nil, so they are equal.
-		return true
+
+	switch x.Type().Kind() {
+	case reflect.Bool:
+		return x.Bool() == y.Bool()
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return x.Int() == y.Int()
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return x.Uint() == y.Uint()
+	case reflect.Float32, reflect.Float64:
+		return x.Float() == y.Float()
+	case reflect.Complex64, reflect.Complex128:
+		return real(x.Complex()) == real(y.Complex()) &&
+			imag(x.Complex()) == imag(y.Complex())
+	case reflect.String:
+		return x.String() == y.String()
+	case reflect.Chan, reflect.Ptr, reflect.UnsafePointer:
+		return x.Pointer() == y.Pointer()
+	case reflect.Array:
+		if x.Len() != y.Len() {
+			return false
+		}
+
+		equal := true
+		for i := 0; i < x.Len(); i++ {
+			if !shallowEqual(x.Index(i), y.Index(i)) {
+				equal = false
+			}
+		}
+		return equal
+	case reflect.Struct:
+		var equal bool
+		for i := 0; i < x.NumField(); i++ {
+			if shallowEqual(x.Field(i), y.Field(i)) {
+				equal = true
+			}
+		}
+		return equal
+	default:
+		panic("type '" + x.Type().Kind().String() + "' is not comparable")
```

Comment:
You can simplify this to:
```go
runtimePanic("comparing uncomparable type")
```
This avoids a few heap allocations and also adds the `runtime error: ` prefix. If you really want to show the type kind, you can also add an extra switch case for each type kind, for example:
```go
case reflect.Slice:
    runtimePanic("slice is not comparable")
```

Reply:
I am fine with it not showing. I was just following other's recommendations. I feel like you.